File: questions/c/EulQ02.c
// Even Fibonacci numbers

#include "essentials.c"

long int even_fib_sum(long int max);

int main(void)
{
    printf("Input positive integer: ");
    long int a; scanf("%li", &a);

    timeStart();
    
    printf("Sum of all even Fibonacci numbers under %li is %li\n", a, even_fib_sum(a));
    
    timeEnd();
    return 0;
}

long int even_fib_sum(long int max)
{
    long int lastTerm = 1;
    long int currentTerm = 1;
    long int theoryTerm = 1;
    long int sum = 0;
    
    while (currentTerm < max)
    {
        if ((currentTerm + lastTerm) % 2 == 0)
        {
            sum += currentTerm + lastTerm;
        }
        theoryTerm = lastTerm;
        lastTerm = currentTerm;
        currentTerm += theoryTerm;
    }
    
    return sum;
}

File: questions/python/EulQ02.py
# Even Fibonacci numbers

import timeit as t
from itertools import takewhile

def fib():
    a = 0
    b = 1
    c = 0
    while True:
        yield a
        a = b
        b = c + b
        c = a

def even_fib_sum(x):
    fibs = filter(lambda x: x % 2 == 0, fib())
    return sum(list(takewhile(lambda n: n < x, fibs)))

a = int(input())

start = t.default_timer()
print(even_fib_sum(a))

end = t.default_timer()
print("Time elapsed:", end-start)